### ISSUE TITLE
[RFC] Reduce frozen Wasm binary size

### ISSUE THREAD

**Issue by lastmjs** [RustPython/RustPython#4203]
## Summary

As I've been building [Kybra](https://github.com/demergent-labs/kybra) (a Python environment for a decentralized cloud), I've run into issues with the final Wasm binary produced when including the `stdlib` and the `frozen stdlib`. The Wasm environment we are deploying to has very strict requirements for Wasm binary size, which right now is ~2mb. That limit will hopefully be lifted over time. But right now RustPython will create Wasm binaries up to ~30mb in size unoptimized, and even optimizing and compressing with gzip results in binaries far too large to deploy to the IC.

I suggest RustPython implement means to reduce the size of the final binary produced by incorporating RustPython. This could be done in a number of ways, including module bundling, tree shaking, dead code elimination, and perhaps other ways as well.

For module bundling, the idea is to follow the statically defined Python imports from an entry file, and create an output directory that only includes the exact files/folders referenced from the imports. The JS ecosystem builders/bundlers make heavy use of this kind of static analysis to create JS bundles. Going beyond this, tree shaking or dead code elimination could be applied to remove unnecessary code.

Here are some helpful libraries that I've researched or even experimented with to help solve some of these problems:

- Module bundling: https://github.com/ronaldoussoren/modulegraph
- Tree shaking: https://github.com/sclabs/treeshaker
- Dead code elimination:

**Comment by youknowone** [comment]
I agree we'd better to invest to this point. Current RustPython binary size is not very fit for common wasm usage.

**Comment by lastmjs** [comment]
So I've got some data on binary sizes to share. I've tested out just a couple examples. What's shown below is the before and after binary size when adding `vm.add_frozen(rustpython_pylib::frozen_stdlib());` inside `rustpython_vm::Interpreter::with_init`.

## Project 1

Wasm binary size before: 8.2 MiB
Wasm binary size after: 19.7 MiB

## Project 2

Wasm binary size before: 5.5 MiB
Wasm binary size after: 16.9 MiB

The conclusion from this simple experiment shows an increase of ~11.5 MiB
